### PR TITLE
fix leaks on level-zero interop program and kernel handles

### PR DESCRIPTION
**Pull request by pbalcer**

### CODE REVIEW

**Comment by pbalcer** [comment]
@nrspruit having looked at this code, device -> handle maps for both kernel and programs may have introduced quite a few subtle issues that are hard to detect. This stems from state duplication. We should probably refactor this code so that the native handles are ideally stored only in one place.

File: test/adapters/level_zero/level_zero_memcheck.match [hunk 1]
```diff
@@ -0,0 +1,2 @@
+{{IGNORE}}
+{{.*}} ERROR SUMMARY: {{[03]}} errors from {{[03]}} contexts {{.*}}
```
Comment (pbalcer) [inline test/adapters/level_zero/level_zero_memcheck.match:2]: I was forced to allow 3 leaks. They originate in libigc, so there's not much we can do here. This should go away once we update all our runners to a newer driver version.

**Comment by pbalcer** [comment]
SYCL E2E failure is unrelated. It's failing on other PRs as well (e.g., https://github.com/intel/llvm/actions/runs/9206096279/job/25325394817?pr=13892).

**Review by igchor — APPROVED** [review]

**Comment by igchor** [comment]
> @nrspruit having looked at this code, device -> handle maps for both kernel and programs may have introduced quite a few subtle issues that are hard to detect. This stems from state duplication. We should probably refactor this code so that the native handles are ideally stored only in one place.

I started doing some refactor here: https://github.com/oneapi-src/unified-runtime/pull/1607/files

There's also a problem with handling multiple zeKernels that I want to address. It would be good to have a machine where we can test this with multiple devices, tough.

**Comment by pbalcer** [comment]
All the test failures are unrelated.

**Comment by pbalcer** [comment]
> It would be good to have a machine where we can test this with multiple devices

No promises, but hopefully we can add a runner with multiple devices next week.

**Pull request merged**